### ISSUE TITLE
severe: export_repair_missing_directory

### ISSUE THREAD

**Issue by lbeltrame** [rozofs/rozofs#146]
When trying to debug some issues in my RozoFS debug (existing files being reported as "no such file or directory") I came across this in the export server's logs:

```
export_tracking.c - 581 - severe: export_repair_missing_directory(/srv/rozofs/exports/NGS/161/00000000-0000-4000-a100-000000009a1f) success (success=79,failure=0)
```

Indeed, that folder does not exist. The logs contain similar other errors. What could be the cause? I wonder if it's related to the issues I'm facing.

**Comment by rozodopamine** [comment]
Could you please run rozo_dumpcnf on the node where the export is actually
running and send
me back the result file.
Regards,
JPierre

Jean-Pierre Monchanin - Customer support


*jean-pierre.monchanin@rozofs.com
<jean-pierre.monchanin@rozofs.com>www.rozofs.com
<http://www.rozofs.com>https://github.com/rozofs/rozofs
<https://github.com/rozofs/rozofs>*



2017-02-09 18:07 GMT+01:00 Luca Beltrame <notifications@github.com>:

**Comment by lbeltrame** [comment]
Thanks Jean-Pierre, I have sent you via email the output.

**Comment by rozodopamine** [comment]
Hi Luca,
looking at the dump you gave to me I can see that your exportd has crashed.
The cause of the crash is your glibc
that is compiled with some options that controls that the socket numbers
are under 1024. RozoFS can use socket
numbers up to 4096 and it works fine (as long as the glibc do not crash).
You should update your glibc not to crash
for this purpose.
At the moment of the crash some directories where being created. These
directories are referenced in their parent
direntries but their own direntries have not yet been saved before the
crash. This inconsistency is what the following
syslog tells
Feb 10 08:58:05 santabarbara export_slave_1[27391]: export_tracking.c - 581
- severe:
export_repair_missing_directory(/srv/rozofs/exports/NGS/98/00000000-0000-4000-6200-00000000cc1a)
success (success=1,failure=0)
The inconsistency is discovered and fixed,
Please fix your glibc.
Regards,
JPierre


Jean-Pierre Monchanin - Customer support


*jean-pierre.monchanin@rozofs.com
<jean-pierre.monchanin@rozofs.com>www.rozofs.com
<http://www.rozofs.com>https://github.com/rozofs/rozofs
<https://github.com/rozofs/rozofs>*



2017-02-10 9:17 GMT+01:00 Luca Beltrame <notifications@github.com>:

**Comment by lbeltrame** [comment]
Would you be able to point to where in glibc this is defined? "Fixing glibc" on a LTS distribution (such as the one running my systems) may require filing a bug report to the distribution itself.

**Comment by rozodopamine** [comment]
The problem comes from run time checks added in glibc on FD_SET and FD_CLR
called __fdelt_chk().
Compiling glibc without FORTIFY_SOURCE option should remove these checks.
Here are some comments on this:
https://sourceware.org/ml/libc-alpha/2013-05/msg00144.html
Regards,
JPierre

Jean-Pierre Monchanin - Customer support


*jean-pierre.monchanin@rozofs.com
<jean-pierre.monchanin@rozofs.com>www.rozofs.com
<http://www.rozofs.com>https://github.com/rozofs/rozofs
<https://github.com/rozofs/rozofs>*



2017-02-13 9:00 GMT+01:00 Luca Beltrame <notifications@github.com>:

**Comment by lbeltrame** [comment]
Wouldn't it be better to build the packages (since I'm using them from the RozoFS repository) with ` -D_FORTIFY_SOURCE=0` as CFLAGS instead of weakening the whole glibc?

**Comment by rozodopamine** [comment]
You are right, removing all the controls is probably not the safest thing
to do. It looks like the safest fix would be for RozoFS to redefine
its own macro for FD_SET, FD_CLR, FD_ZERO and FD_ISSET in order to bypass
the glibc controls and keep FORTIFY_SOURCE=2 as
compilation option.
We will think about it and a hotfix will probably be done for this.
Regards,
JPierre

Jean-Pierre Monchanin - Customer support


*jean-pierre.monchanin@rozofs.com
<jean-pierre.monchanin@rozofs.com>www.rozofs.com
<http://www.rozofs.com>https://github.com/rozofs/rozofs
<https://github.com/rozofs/rozofs>*



2017-02-13 14:32 GMT+01:00 Luca Beltrame <notifications@github.com>:

**Comment by lbeltrame** [comment]
Thanks. I'll keep this issue open and close when a possible hotfix lands.

**Comment by rozodopamine** [comment]
Fixed in release 2.1.19
